torch/csrc/jit/passes/integer_value_refinement.cpp

```diff
@@ -0,0 +1,215 @@
+// in an example like:
+// if x == 1:
+//    ...
+// we can substitute all uses of x dominated by the true block
+// with 1.
+
+struct IntegerValueRefiner {
+  IntegerValueRefiner(std::shared_ptr<Graph> graph)
+      : graph_(std::move(graph)) {}
+
+  bool run() {
+    if (!checkForPossibleRefinements(graph_->block())) {
+      return false;
+    }
+    IntegerRefinement refinements;
+    RefineIntegerValues(graph_->block(), refinements);
+    return changed_;
+  }
+
+  bool checkForPossibleRefinements(Block* b) {
+    for (Node* n : b->nodes()) {
+      if (n->matches("aten::eq(int a, int b) -> bool") ||
+          n->matches("aten::ne(int a, int b) -> bool")) {
+        for (size_t const_index : {0, 1}) {
+          auto non_const_index = 1 - const_index;
+          if (n->inputs().at(const_index)->node()->kind() == prim::Constant &&
+              n->inputs().at(non_const_index)->uses().size() > 1) {
+            return true;
+          }
+        }
+      }
+      for (Block* block : n->blocks()) {
+        if (checkForPossibleRefinements(block)) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  void removeIfNodeOutputsWithRefinements(
+      Node* if_node,
+      IntegerRefinement& true_block_refinements,
+      IntegerRefinement& false_block_refinements) {
+    // we are looking for cases where we can replace
+    // both block outputs with the same value, which opens up
+    // further optimization opportunities
+    // The pass will already handle if both are refined to the same
+    // constant. Here, we add a case case where one block output
+    // is refined to a constant in the other block, and where the existing
+    // block output in the other block is the same constant
+    // x = 1
+    // if y == 1:
+    //    return x
+    // else:
+    //    return y
+    // can always safely be replaced with `y`
+    // this is an important case for symbolic shape analysis
+
+    for (size_t block_index : {0, 1}) {
+      Block* if_block = if_node->blocks().at(block_index);
+      Block* other_if_block = if_node->blocks().at(1 - block_index);
+      for (size_t i = 0; i < if_node->outputs().size(); ++i) {
+        Value* block_output = if_block->outputs().at(i);
+        if (!block_output->type()->cast<IntType>()) {
+          continue;
+        }
+        // Value must be in scope for both blocks
+        if (!if_node->isDominatedBy(block_output->node())) {
+          continue;
+        }
+        // one constant value one not
+        auto other_const_value =
+            constant_as<int64_t>(other_if_block->outputs().at(i));
+        if (!other_const_value ||
+            block_output->node()->kind() == prim::Constant) {
+          continue;
+        }
+        const auto& other_block_refinements =
+            block_index == 0 ? false_block_refinements : true_block_refinements;
+        c10::optional<int64_t> maybe_refine = tryFindRefinement(block_output);
```

Comment:
This code in particular is very hard to reason about without an example at hand. I would recommend adding example + adding comments explaining what each value in this code is holding in that example. E.g.:
```
block_output = %y.1
maybe_refine = {%x.1 --> 1} // the actual value should probably be different here
other_block_refinements = {%y.1 --> 1}
```

Reply:
how is this different than the example given above ?

Reply:
The example above doesn't show what each of these variables hold (other_block_refinements, block_output, ...) - what I want here is almost a walkthrough the code reflected in the comments, e.g. "after this line 'block_output' holds this", "after that line maybe_refine gets this value".

Also, I'd like to see the example formulated in terms of TorchScript so that it's easier to follow (as I mentioned in the previous comment).